File: problems/84/solution.go
package main
const MaxInt = int(^uint(0) >> 1)
const MinInt = -MaxInt - 1

func largestRectangleArea(heights []int) int {
    n, ans := len(heights), 0
    
    for i := 0; i < n; i++ {
        minHeight := MaxInt
        
        for j := i; j < n; j++ {
            minHeight = min(minHeight, heights[j])
            ans = max(ans, minHeight * (j - i + 1))
        }
    }
    
    return ans
}

func min(a, b int) int {
    if a < b {
        return a
    }
    
    return b
}

func max(a, b int) int {
    if a > b {
        return a
    }
    
    return b
}

File: problems/84/solution.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int ans = 0;
        stack<int> st;
        
        st.push(-1);
        
        for (int i = 0; i < heights.size(); i++) {
            while (st.top() != -1 and heights[st.top()] >= heights[i]) {
                int curHeight = heights[st.top()];
                st.pop();
                
                ans = max(ans, curHeight * (i - st.top() - 1));
            }
            
            st.push(i);
        }
        
        while (st.top() != -1) {
            int curHeight = heights[st.top()];
            st.pop();
            int curWidth = heights.size() - st.top() - 1;
            
            ans = max(ans, curHeight * curWidth);
        }
        
        return ans;
    }
};